File: project_euler/Problem060/Problem060.java
package project_euler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem060 {
    public void solution() {
        int target = 5;
        List<Integer> primes = new ArrayList<>();
        for (int i = 3; i < 10000; i += 2) {
            if (isPrime(i)) {
                primes.add(i);
            }
        }

        int length = primes.size();
        int[][] edges = new int[length][length];
        for (int i = 0; i < length; ++i) {
            Arrays.fill(edges[i], 0);
        }
        for (int i = 1; i < length; ++i) {
            int value = primes.get(i);
            if (checkPrime(merge(3, value), primes) && checkPrime(merge(value, 3), primes)) {
                edges[0][i] = 1;
                edges[i][0] = 1;
            }
        }
        for (int i = 1; i < length; ++i) {
            int pivot = primes.get(i);
            for (int j = i + 1; j < length; ++j) {
                int value = primes.get(j);
                if (pivot % 3 == value % 3 && checkPrime(merge(pivot, value), primes) && checkPrime(merge(value, pivot), primes)) {
                    edges[i][j] = 1;
                    edges[j][i] = 1;
                }
            }
        }

        List<Integer> cliques = new ArrayList<>();
        List<Integer> indexes = new ArrayList<>();
        for (int i = 0; i < length; ++i) {
            indexes.add(i);
        }
        search(primes, edges, cliques, indexes, target);
    }

    private boolean isPrime(int value) {
        if (0 == (value & 1)) {
            return 2 == value;
        }

        int border = (int) Math.sqrt(value);
        for (int i = 3; i <= border; i += 2) {
            if (0 == value % i) {
                return false;
            }
        }
        return 1 != value;
    }

    private boolean checkPrime(int value, List<Integer> primes) {
        if (value < 10000) {
            int start = 0;
            int end = primes.size() - 1;
            while (start <= end) {
                int current = start + ((end - start) >> 1);
                int prime = primes.get(current);
                if (prime == value) {
                    return true;
                }

                if (prime < value) {
                    start = current + 1;
                } else {
                    end = current - 1;
                }
            }
            return false;
        }

        int border = (int) Math.sqrt(value);
        int start = 0;
        int end = primes.size() - 1;
        while (start <= end) {
            int current = start + ((end - start) >> 1);
            if (primes.get(current) <= border) {
                start = current + 1;
            } else {
                end = current - 1;
            }
        }

        for (int i = 0; i < start; ++i) {
            if (0 == value % primes.get(i)) {
                return false;
            }
        }
        return true;
    }

    private int merge(int head, int tail) {
        return head * (int) Math.pow(10, (int) Math.log10(tail) + 1) + tail;
    }

    private void search(List<Integer> primes, int[][] edges, List<Integer> cliques, List<Integer> indexes, int target) {
        if (indexes.isEmpty()) {
            if (cliques.size() == target) {
                int result = 0;
                for (int index : cliques) {
                    result += primes.get(index);
                }
                System.out.println(result);
            }
            return;
        }

        int start = indexes.get(0);
        for (int i = indexes.size() - 1; i >= 0; --i) {
            int index = indexes.get(i);
            if (0 == edges[start][index]) {
                cliques.add(index);
                search(primes, edges, cliques, getAdjacents(edges, indexes, index), target);
                cliques.remove(cliques.size() - 1);
                indexes.remove((Integer) index);
            }
        }
    }

    private List<Integer> getAdjacents(int[][] edges, List<Integer> indexes, int start) {
        List<Integer> auxiliary = new ArrayList<>();
        for (int index : indexes) {
            if (0 != edges[start][index]) {
                auxiliary.add(index);
            }
        }
        return auxiliary;
    }
}

File: project_euler/Problem060/Problem060.py
import math
from typing import List

class Problem060:
    def solution(self) -> None:
        target, primes = 5, []
        for i in range(3, 10000, 2):
            if self.__is_prime(i):
                primes.append(i)

        length = len(primes)
        edges = [[0 for _ in range(length)] for _ in range(length)]
        for i in range(1, length):
            value = primes[i]
            if self.__check_prime(self.__merge(3, value), primes) and self.__check_prime(self.__merge(value, 3), primes):
                edges[0][i], edges[i][0] = 1, 1
        for i in range(1, length):
            pivot = primes[i]
            for j in range(i + 1, length):
                value = primes[j]
                if pivot % 3 == value % 3 and self.__check_prime(self.__merge(pivot, value), primes) and self.__check_prime(self.__merge(value, pivot), primes):
                    edges[i][j], edges[j][i] = 1, 1

        cliques, indexes = [], []
        for i in range(length):
            indexes.append(i)
        self.__search(primes, edges, cliques, indexes, target)

    def __is_prime(self, value: int) -> bool:
        if 1 != value % 6 and 5 != value % 6:
            return 2 == value or 3 == value

        border = int(math.sqrt(value))
        for i in range(5, border + 1, 6):
            if not value % i or not value % (i + 2):
                return False
        return 1 != value

    def __check_prime(self, value: int, primes: List[int]) -> bool:
        if value < 10000:
            start, end = 0, len(primes) - 1
            while start <= end:
                current = start + ((end - start) >> 1)
                prime = primes[current]
                if prime == value:
                    return True

                if prime < value:
                    start = current + 1
                else:
                    end = current - 1
            return False

        border, start, end = int(math.sqrt(value)), 0, len(primes) - 1
        while start <= end:
            current = start + ((end - start) >> 1)
            if primes[current] <= border:
                start = current + 1
            else:
                end = current - 1

        for prime in primes[:start]:
            if not value % prime:
                return False
        return True

    def __merge(self, head: int, tail: int) -> int:
        return head * 10 ** (int(math.log10(tail)) + 1) + tail

    def __search(self, primes: List[int], edges: List[List[int]], cliques: List[int], indexes: List[int], target: int) -> None:
        if not indexes:
            if len(cliques) == target:
                result = 0
                for index in cliques:
                    result += primes[index]
                print(result)
            return

        start = indexes[0]
        for i in range(len(indexes) - 1, -1, -1):
            index = indexes[i]
            if not edges[start][index]:
                cliques.append(index)
                self.__search(primes, edges, cliques, self.__get_adjacents(edges, indexes, index), target)
                cliques.pop()
                indexes.remove(index)

    def __get_adjacents(self, edges: List[List[int]], indexes: List[int], start: int) -> List[int]:
        auxiliary = []
        for index in indexes:
            if edges[start][index]:
                auxiliary.append(index)
        return auxiliary
